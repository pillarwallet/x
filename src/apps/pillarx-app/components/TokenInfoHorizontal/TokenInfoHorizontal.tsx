import { useEffect, useState } from 'react';

// types
import { TokenContract } from '../../../../types/api';

// reducer
import { useGetBlockchainsListQuery } from '../../../token-atlas/api/token';

// components
import RandomAvatar from '../RandomAvatar/RandomAvatar';
import TokensPercentage from '../TokensPercentage/TokensPercentage';
import Body from '../Typography/Body';
import BodySmall from '../Typography/BodySmall';

type TokenInfoHorizontalProps = {
  logo?: string;
  tokenName?: string;
  tokenValue?: number;
  percentage?: number;
  tokenChains?: TokenContract[];
  onClick?: () => void;
};

const TokenInfoHorizontal = ({
  logo,
  tokenName,
  tokenValue,
  percentage,
  tokenChains,
  onClick,
}: TokenInfoHorizontalProps) => {
  const [blockchainLogo, setBlockchainLogo] = useState<string | undefined>(
    undefined
  );
  const [isBrokenImage, setIsBrokenImage] = useState<boolean>(false);

  const { data: blockchainListData, isSuccess } = useGetBlockchainsListQuery();

  useEffect(() => {
    if (tokenChains?.length === 1 && blockchainListData && isSuccess) {
      const chainData = blockchainListData.data.find(
        (chain) => chain.name === tokenChains[0].blockchain
      );
      setBlockchainLogo(chainData?.logo);
    }
  }, [blockchainListData, isSuccess, tokenChains]);

  return (
    <div
      className="flex flex-col py-5 px-[22px] gap-1 w-[122px] h-auto items-center cursor-pointer tablet:w-[120px] mobile:w-[100px] mobile:px-3.5"
      onClick={onClick}
    >
      <div className="relative w-[70px] h-[70px] rounded-full mb-3.5">
        {logo && !isBrokenImage ? (
          <img
            src={logo}
            alt="token-logo"
            className="w-full h-full object-fill rounded-full"
            data-testid="token-info-horizontal-logo"
            onError={() => setIsBrokenImage(true)}
          />
        ) : (
          <div className="w-full h-full overflow-hidden rounded-full">
            <RandomAvatar name={tokenName || ''} />
          </div>
        )}

        {/* Overlay text when no logo available */}
        {(!logo || isBrokenImage) && (
          <span className="absolute inset-0 flex items-center justify-center text-white text-lg font-bold">
            {tokenName?.slice(0, 2)}
          </span>
        )}

        {/* Blockchain logo overlapping when only one blockchain for this token */}
        {tokenChains?.length === 1 ? (
          <div className="absolute bottom-0 right-0 w-6 h-6 rounded-full overflow-hidden border-[1px] bg-white border-container_grey transform translate-x-1/5 translate-y-1/5">
            <img
              src={blockchainLogo}
              alt="logo"
              className="w-full h-full object-contain"
            />
          </div>
        ) : null}
      </div>

      {tokenName && <Body className="text-center">{tokenName}</Body>}
      {tokenValue && (
        <BodySmall className="text-center">${tokenValue.toFixed(4)}</BodySmall>
      )}
      <TokensPercentage percentage={percentage} />
    </div>
  );
};

export default TokenInfoHorizontal;
