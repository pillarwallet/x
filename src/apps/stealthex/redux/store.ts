import { configureStore, TypedStartListening } from '@reduxjs/toolkit';
import exchangeReducer from './reducers/exchange';
import { exchange } from '../lib/backend/api';

export const store = configureStore({
    reducer: {
        exchange: exchangeReducer,
        [exchange.reducerPath]: exchange.reducer
    },
    middleware: getDefaultMiddleware =>
        getDefaultMiddleware({ serializableCheck: false }).concat(exchange.middleware),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
export type AppStartListening = TypedStartListening<RootState, AppDispatch>; 