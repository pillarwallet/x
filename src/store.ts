/* eslint-disable @typescript-eslint/no-use-before-define */
// Core
import {
  Middleware,
  Reducer,
  combineReducers,
  configureStore,
  createDynamicMiddleware,
} from '@reduxjs/toolkit';
import { setupListeners } from '@reduxjs/toolkit/query';

// Services
import swapSlice from './apps/the-exchange/reducer/theExchangeSlice';
import tokenAtlasSlice from './apps/token-atlas/reducer/tokenAtlasSlice';
import { pillarXApiPresence } from './services/pillarXApiPresence';
import { pillarXApiTransactionsHistory } from './services/pillarXApiTransactionsHistory';
import { pillarXApiWaitlist } from './services/pillarXApiWaitlist';

// Initialisation
const dynamicMiddleware = createDynamicMiddleware();
const middlewareReducers: { [key: string]: Reducer } = {};

/**
 * @name addReducer
 * @description addReducer is a function that allows anyone
 * working with PillarX to add reducers to the store
 * from their own application.
 *
 * @param newReducer
 */
export const addReducer = (newReducer: {
  reducerPath: string;
  reducer: Reducer;
}) => {
  middlewareReducers[newReducer.reducerPath as string] = newReducer.reducer;
  store.replaceReducer(combineReducers(middlewareReducers));
};

/**
 * @name addMiddleware
 * @description addMiddleware is a function that allows anyone
 * working with PillarX to add RTK middleware to the store
 * from their own application.
 *
 * @param newMiddleware
 */
export const addMiddleware = (newMiddleware: {
  reducerPath: string;
  reducer: Reducer;
  middleware: Middleware;
}) => {
  middlewareReducers[newMiddleware.reducerPath as string] =
    newMiddleware.reducer;
  dynamicMiddleware.addMiddleware(newMiddleware.middleware);

  store.replaceReducer(combineReducers(middlewareReducers));
};

/**
 * Export a store from RTK
 */
export const store = configureStore({
  // Empty reducer for now - the addMiddleware function
  // below will dynamically regenerate the reducers required
  // from the middleware functions.
  reducer: {},
  // Adding the api middleware enables caching, invalidation, polling,
  // and other useful features of `rtk-query`.
  // Note: here we have added dynamicMiddleware.middleware
  // which instructs the store to use the middleware that we
  // will add later.
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().prepend(dynamicMiddleware.middleware),
});

/**
 * Whilst we're here we'll immediately mount the
 * PillarXApi middleware. This will contain all the
 * core API functionality for the PillarX application.
 */
addMiddleware(pillarXApiWaitlist);
addMiddleware(pillarXApiPresence);
addMiddleware(pillarXApiTransactionsHistory);
addReducer(swapSlice);
addReducer(tokenAtlasSlice);

// optional, but required for refetchOnFocus/refetchOnReconnect behaviors
// see `setupListeners` docs - takes an optional callback as the 2nd arg for customization
setupListeners(store.dispatch);

// // Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;
