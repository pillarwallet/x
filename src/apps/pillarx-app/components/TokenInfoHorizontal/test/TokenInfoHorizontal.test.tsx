import { render, screen } from '@testing-library/react';
import renderer, { ReactTestRendererJSON } from 'react-test-renderer';

// provider
import { Provider } from 'react-redux';
import { store } from '../../../../../store';

// components
import TokenInfoHorizontal from '../TokenInfoHorizontal';

describe('<TokenInfoHorizontal />', () => {
  const logo = 'https://example.com/logo.png';
  const tokenName = 'Token Name';
  const tokenValue = 1000;
  const percentage = 5.45;

  it('renders correctly and matches snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <TokenInfoHorizontal
            logo={logo}
            tokenName={tokenName}
            tokenValue={tokenValue}
            percentage={percentage}
          />
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders the logo', () => {
    render(
      <Provider store={store}>
        <TokenInfoHorizontal
          logo={logo}
          tokenName={tokenName}
          tokenValue={tokenValue}
          percentage={percentage}
        />
      </Provider>
    );

    expect(
      screen.getByTestId('token-info-horizontal-logo')
    ).toBeInTheDocument();
    expect(screen.queryByTestId('random-avatar')).not.toBeInTheDocument();
  });

  it('renders the default logo (random avatar) when logo is not provided', () => {
    render(
      <Provider store={store}>
        <TokenInfoHorizontal
          tokenName={tokenName}
          tokenValue={tokenValue}
          percentage={percentage}
        />
      </Provider>
    );

    expect(screen.getByTestId('random-avatar')).toBeInTheDocument();
    expect(
      screen.queryByTestId('token-info-horizontal-logo')
    ).not.toBeInTheDocument();
  });

  it('renders the token name correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <TokenInfoHorizontal
            logo={logo}
            tokenName={tokenName}
            tokenValue={tokenValue}
            percentage={percentage}
          />
        </Provider>
      )
      .toJSON() as ReactTestRendererJSON;
    const tokenNameProp = tree.children?.find(
      (child) => typeof child === 'object' && child.type === 'p'
    ) as ReactTestRendererJSON;

    expect(tokenNameProp).not.toBeNull();
    expect(tokenNameProp.type).toBe('p');
    expect(tokenNameProp.children).toContain(tokenName);
  });

  it('does not render the token name', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <TokenInfoHorizontal
            logo={logo}
            tokenValue={tokenValue}
            percentage={percentage}
          />
        </Provider>
      )
      .toJSON() as ReactTestRendererJSON;
    const tokenNameProp = tree.children?.find(
      (child) => typeof child === 'object' && child.type === 'p'
    ) as ReactTestRendererJSON;

    expect(tokenNameProp.children).not.toContain(tokenName);
  });

  it('renders the token value correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <TokenInfoHorizontal
            logo={logo}
            tokenName={tokenName}
            tokenValue={tokenValue}
            percentage={percentage}
          />
        </Provider>
      )
      .toJSON() as ReactTestRendererJSON;

    const tokenValueProp = tree.children?.filter(
      (child) => typeof child === 'object' && child.type === 'p'
    )[1] as ReactTestRendererJSON;

    expect(tokenValueProp).not.toBeNull();
    expect(tokenValueProp.type).toBe('p');
    expect(tokenValueProp.children).toContain(tokenValue.toFixed(4));
  });

  it('does not render the token value', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <TokenInfoHorizontal
            logo={logo}
            tokenName={tokenName}
            percentage={percentage}
          />
        </Provider>
      )
      .toJSON() as ReactTestRendererJSON;

    const tokenValueProp = tree.children?.find(
      (child) => typeof child === 'object' && child.type === 'p'
    ) as ReactTestRendererJSON;

    expect(tokenValueProp.children).not.toContain(tokenValue.toFixed(4));
  });

  it('applies the correct style', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <TokenInfoHorizontal
            logo={logo}
            tokenName={tokenName}
            tokenValue={tokenValue}
            percentage={percentage}
          />
        </Provider>
      )
      .toJSON() as ReactTestRendererJSON;
    expect(tree.type).toBe('div');
    expect(tree.props.className).toContain('flex');
    expect(tree.props.className).toContain('flex-col');
    expect(tree.props.className).toContain('h-auto');
    expect(tree.props.className).toContain('py-5');
  });
});
