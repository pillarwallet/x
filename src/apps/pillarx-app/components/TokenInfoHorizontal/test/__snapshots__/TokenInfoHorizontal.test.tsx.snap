// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<TokenInfoHorizontal /> > renders correctly and matches snapshot 1`] = `
<div
  className="flex flex-col py-5 px-[22px] gap-1 w-[122px] h-auto items-center cursor-pointer tablet:w-[120px] mobile:w-[100px] mobile:px-3.5"
>
  <div
    className="relative w-[70px] h-[70px] rounded-full mb-3.5"
  >
    <img
      alt="token-logo"
      className="w-full h-full object-fill rounded-full"
      data-testid="token-info-horizontal-logo"
      onError={[Function]}
      src="https://example.com/logo.png"
    />
  </div>
  <p
    className="text-base font-medium text-center"
  >
    Token Name
  </p>
  <p
    className="text-sm font-medium text-center"
  >
    $
    1000.0000
  </p>
  <div
    className="flex gap-1 items-center"
  >
    <img
      alt="percentage-up"
      className="w-[10px] h[9px]"
      src="data:image/svg+xml,%3csvg%20width='10'%20height='10'%20viewBox='0%200%2010%2010'%20fill='none'%20xmlns='http://www.w3.org/2000/svg'%3e%3cpath%20d='M4.03034%201.08115C4.4613%200.334701%205.5387%200.334701%205.96967%201.08115L9.84832%207.79918C10.2793%208.54563%209.74058%209.47869%208.87866%209.47869L5%209.47869L1.12134%209.47869C0.259418%209.47869%20-0.279283%208.54563%200.151678%207.79918L4.03034%201.08115Z'%20fill='%2305FFDD'/%3e%3c/svg%3e"
    />
    <p
      className="text-sm font-medium text-percentage_green"
    >
      5.45
      %
    </p>
  </div>
</div>
`;
